src/hooks/core/users/useFollowUserV2Hook.tsx

```diff
@@ -0,0 +1,135 @@
+
+    try {
+      const result = await checkFollowStatus(channelId);
+      StorageUtils.otherProfileData.setForKey(channelId, JSON.stringify(result));
+    } catch (e) {
+      console.error('get follow status failed', e);
+    }
+  }
+
+  const getOtherProfileFollowStatus = async (channelId: string | undefined): Promise<any> => {
+    if (!channelId) return;
+
+    let otherProfileDataCache = StorageUtils.otherProfileData.getForKey(channelId);
+    let otherProfileDataJSON: FollowStatusAPI = {
+      isMeFollowingTargetFromAction: undefined,
+      isMeFollowingTarget: undefined,
+      isTargetFollowingMe: undefined
+    };
+
+    try {
+      otherProfileDataJSON = JSON.parse(otherProfileDataCache || '{}');
+    } catch (e) {
+      console.error(e);
+    }
+
+    if (otherProfileDataCache) {
+      setFollowStatus({
+        isFollowed: otherProfileDataJSON?.isTargetFollowingMe,
+        isFollowing: otherProfileDataJSON?.isMeFollowingTarget,
+        isFollowingFromAction: otherProfileDataJSON?.isMeFollowingTargetFromAction
+      });
+    }
+
+    await __helperCallOtherProfileData(channelId);
+    otherProfileDataCache = StorageUtils.otherProfileData.getForKey(channelId);
+    try {
+      otherProfileDataJSON = JSON.parse(otherProfileDataCache || '{}');
+    } catch (e) {
+      console.error(e);
+    }
+
+    setFollowStatus({
+      isFollowed: otherProfileDataJSON?.isTargetFollowingMe,
+      isFollowing: otherProfileDataJSON?.isMeFollowingTarget,
+      isFollowingFromAction: otherProfileDataJSON?.isMeFollowingTargetFromAction
+    });
+  };
+
+  const followUserAction = async (item: ChannelList) => {
+    /**
+     * Optimistic state update (START)
+     */
+    const initialFollowStatus = {...followStatus};
+    setFollowStatus((prevFollowStatus) => ({
+      ...prevFollowStatus,
+      isFollowingFromAction: !initialFollowStatus?.isFollowingFromAction,
+      isFollowing: !initialFollowStatus?.isFollowing
+    }));
+
+    /**
+     * Optimistic state update (END)
+     */
+
+    try {
+      const myId = tab === 'signed' ? signedProfileId : anonProfileId;
+      const targetUser = item?.rawJson?.channel?.members?.find(
+        (member) => member?.user_id !== myId
+      )?.user;
+
+      const data = {
+        user_id_followed: targetUser?.id,
+        follow_source: 'chat'
+      };
+
+      if (initialFollowStatus?.isFollowing) {
+        await setUnFollow(data);
+      } else {
+        await setFollow(data);
+      }
+      __helperCallOtherProfileData(item?.id);
+    } catch (e) {
+      const action = initialFollowStatus?.isFollowing ? 'unfollow' : 'follow';
+      SimpleToast.show(`Failed to ${action} user. Please try again later`);
+      /**
+       * Revert optimistic update
+       */
+      setFollowStatus(() => ({
+        ...initialFollowStatus
+      }));
+    }
+  };
+
+  return {
+    followStatus,
+
+    followUserAction,
+    getOtherProfileFollowStatus
+  };
+};
+
+export default useFollowUserV2Hook;
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The new hook `useFollowUserV2Hook` seems to be well implemented. It provides a clear separation of concerns, with each function handling a specific task related to the follow functionality. The use of optimistic updates is a good practice for improving user experience by providing immediate feedback.

However, there are some areas that could be improved:

1. Error Handling: Currently, the error handling is limited to console logging and showing a toast message. It would be beneficial to have more robust error handling, perhaps by returning an error state from the hook that can be used by the components to handle errors in a way that fits their context.

2. Caching: The hook uses local storage to cache the follow status. This is a good practice for reducing unnecessary network requests. However, it's important to ensure that the cache is invalidated when necessary to prevent stale data from being used.

3. Async/Await: The async/await pattern is used correctly throughout the hook. However, it's important to note that any consumer of this hook must also use async/await or Promises to handle the asynchronous operations correctly.

4. TypeScript: The TypeScript types are well defined and provide a clear understanding of the data structures used by the hook. However, the `FollowStatusAPI` type has properties that are optional and can be undefined. If these properties are always expected to be present (even if they may be false), consider making them non-optional to enforce stricter type checking.

5. Code Duplication: There is some code duplication in the `getOtherProfileFollowStatus` function where the follow status is set based on the cached data. Consider refactoring this into a separate function to adhere to the DRY principle.

Here's a suggestion to address the code duplication issue:

```diff
  const setFollowStatusFromCache = (cache: string) => {
    let otherProfileDataJSON: FollowStatusAPI = {
      isMeFollowingTargetFromAction: undefined,
      isMeFollowingTarget: undefined,
      isTargetFollowingMe: undefined
    };

    try {
      otherProfileDataJSON = JSON.parse(cache || '{}');
    } catch (e) {
      console.error(e);
    }

    setFollowStatus({
      isFollowed: otherProfileDataJSON?.isTargetFollowingMe,
      isFollowing: otherProfileDataJSON?.isMeFollowingTarget,
      isFollowingFromAction: otherProfileDataJSON?.isMeFollowingTargetFromAction
    });
  };

  const getOtherProfileFollowStatus = async (channelId: string | undefined): Promise<any> => {
    if (!channelId) return;

    let otherProfileDataCache = StorageUtils.otherProfileData.getForKey(channelId);

    if (otherProfileDataCache) {
      setFollowStatusFromCache(otherProfileDataCache);
    }

    await __helperCallOtherProfileData(channelId);
    otherProfileDataCache = StorageUtils.otherProfileData.getForKey(channelId);
    setFollowStatusFromCache(otherProfileDataCache);
  };
```

Overall, the new hook `useFollowUserV2Hook` seems to be a good improvement over the previous `useFollowUser` hook. It provides a more modular and maintainable approach to managing follow functionality in the application.


<!-- This is an auto-generated comment by OSS CodeRabbit -->